[web-animations] WPT test css/css-animations/translation-animation-on-important-property.html is a failure

graouts opened the pull request:
#### 50450e8c417a9b25ff80bc6e2bc76e9f238fe475
<pre>
[web-animations] WPT test css/css-animations/translation-animation-on-important-property.html is a failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=252481">https://bugs.webkit.org/show_bug.cgi?id=252481</a>

Reviewed by Antti Koivisto.

WPT.fyi reports we are failing the recently added css/css-animations/translation-animation-on-important-property.html.
This test fails because it is animating an accelerated property, in this case &quot;transform&quot;, but also setting an
!important style for it, checking that the !important value wins over the animated value. We fail to account for the
value set by the !important style when running accelerated animations.

To address this, we add a new method KeyframeEffectStack::acceleratedPropertiesOverriddenByCascade() called by
Style::TreeResolver::createAnimatedElementUpdate() with the animated style prior to further application of the
cascade and after the cascade has been fully applied. In this new method we see which accelerated properties have
a different value between the two styles which informs whether an !important style – or any style with higher
importance than animated style in the cascade - overrides the value set by animations.

We store this list of properties and inform all effects in the stack when this list changes. Then, an effect can determine
whether the presence of such a property prevents it from running accelerated and, using the existing system to interrupt
and resume accelerated animations based on whether other effects in the stack can or cannot be accelerated, will toggle
its accelerated running state.

Since the WPT test this patch fixes does not check whether animations can be interrupted and resumed dynamically
by the presence of an overriding static property, we add such a test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/translation-animation-on-important-property-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/translation-animation-on-important-property.html: Added.
* LayoutTests/webanimations/combining-transform-animations-with-different-acceleration-capabilities-4-expected.txt: Added.
* LayoutTests/webanimations/combining-transform-animations-with-different-acceleration-capabilities-4.html: Added.
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::setBlendingKeyframes):
(WebCore::KeyframeEffect::canBeAccelerated const):
(WebCore::KeyframeEffect::computeHasAcceleratedPropertyOverriddenByCascadeProperty):
(WebCore::KeyframeEffect::acceleratedPropertiesOverriddenByCascadeDidChange):
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::didApplyCascade):
* Source/WebCore/animation/KeyframeEffectStack.h:
(WebCore::KeyframeEffectStack::acceleratedPropertiesOverriddenByCascade const):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):

Canonical link: <a href="https://commits.webkit.org/260464@main">https://commits.webkit.org/260464@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04077f154061cb4ce81c68b9697f205ce7facad1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/6/builds/108407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/17501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/41266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/117519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/112293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/8779 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/100631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/114175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/41266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/100631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/41266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/100631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/10328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/41266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/11074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/8779 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/16477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/41266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/12660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3932 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/animation/KeyframeEffectStack.cpp:
```diff
@@ -247,4 +247,26 @@ void KeyframeEffectStack::lastStyleChangeEventStyleDidChange(const RenderStyle*
         effect->lastStyleChangeEventStyleDidChange(previousStyle, currentStyle);
 }
 
+void KeyframeEffectStack::cascadeWasApplied(const RenderStyle& styleBeforeCascade, const RenderStyle& styleAfterCascade, const HashSet<AnimatableProperty> animatedProperties, const Document& document)
```
`didApplyCascade` might be more inline with the usual naming

graouts replied:
Agreed, will fix when landing.

graouts committed:
[web-animations] WPT test css/css-animations/translation-animation-on-important-property.html is a failure
https://bugs.webkit.org/show_bug.cgi?id=252481

Reviewed by Antti Koivisto.

WPT.fyi reports we are failing the recently added css/css-animations/translation-animation-on-important-property.html.
This test fails because it is animating an accelerated property, in this case "transform", but also setting an
!important style for it, checking that the !important value wins over the animated value. We fail to account for the
value set by the !important style when running accelerated animations.

To address this, we add a new method KeyframeEffectStack::acceleratedPropertiesOverriddenByCascade() called by
Style::TreeResolver::createAnimatedElementUpdate() with the animated style prior to further application of the
cascade and after the cascade has been fully applied. In this new method we see which accelerated properties have
a different value between the two styles which informs whether an !important style – or any style with higher
importance than animated style in the cascade - overrides the value set by animations.

We store this list of properties and inform all effects in the stack when this list changes. Then, an effect can determine
whether the presence of such a property prevents it from running accelerated and, using the existing system to interrupt
and resume accelerated animations based on whether other effects in the stack can or cannot be accelerated, will toggle
its accelerated running state.

Since the WPT test this patch fixes does not check whether animations can be interrupted and resumed dynamically
by the presence of an overriding static property, we add such a test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/translation-animation-on-important-property-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/translation-animation-on-important-property.html: Added.
* LayoutTests/webanimations/combining-transform-animations-with-different-acceleration-capabilities-4-expected.txt: Added.
* LayoutTests/webanimations/combining-transform-animations-with-different-acceleration-capabilities-4.html: Added.
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::setBlendingKeyframes):
(WebCore::KeyframeEffect::canBeAccelerated const):
(WebCore::KeyframeEffect::computeHasAcceleratedPropertyOverriddenByCascadeProperty):
(WebCore::KeyframeEffect::acceleratedPropertiesOverriddenByCascadeDidChange):
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::didApplyCascade):
* Source/WebCore/animation/KeyframeEffectStack.h:
(WebCore::KeyframeEffectStack::acceleratedPropertiesOverriddenByCascade const):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):

Canonical link: https://commits.webkit.org/260464@main
